### DISCUSSION TITLE
How to force JsonResponse/prevent redirect?

### DISCUSSION THREAD

**Discussion by jeremyswright** [spatie/laravel-data#636] [Q&A]
Apologies, but I've only just picked up this library so I've probably just missed something. Hoping someone can point me in the right direction.

Previously when using FormRequests to validate requests coming into API routes, I've provided a Rules function, then provided a failedValidation routine to generate a HttpResponseException, which if the request content type is JSON, returns a JsonResponse.
```
    /**
     * Render any validation errors as a JsonResponse rather than a redirect with flashed error messages
     *
     * @param Validator $validator
     * @return void
     */
    public function failedValidation(Validator $validator)
    {
        $errors = $validator->errors();

        $response = response()->json([
            'message' => 'invalid parameter',
            'details' => $errors->messages(),
        ], Response::HTTP_UNPROCESSABLE_ENTITY);

        throw new HttpResponseException($response);
    }
```
So I created a Data class as follows
```
class OperationLogData extends Data
{
    public function __construct(
        public string $param1,
        public string $param2,
    ) {
    }
}
```
When I submit a request with both parameters, it works fine. When I submit with one of the parameters missing I see a 302 to the homepage.  I'd expect that to instead return a 422 with appropriate message.  How do I force the errors to be reported as a JsonResponse?

Thanks.

**Comment by rubenvanassche** [comment]
Data will only throw an ValidationException which the framework will handle, implementing a thing like `failedValidation` is nearly impossible with data due to nesting and collections. Why not catch the validation exception and throw the new exception? Put it in a helper method? Or create a new abstract data class from which all your data classes extend, then overwrite the `factory` method so that you can catch validation errors and throw the new exception.
